### PR TITLE
add retries to gcs (up/down)loads to make remote logging more resilient

### PR DESCRIPTION
**Pull request by spennihana**

### CODE REVIEW

File: airflow/hooks/base_hook.py [hunk 1]
```diff
@@ -44,6 +44,21 @@ class BaseHook(LoggingMixin):
     def __init__(self, source):
         pass
 
+    @classmethod
+    def _get_connection_from_db_with_retries(cls, conn_id, max_tries=5):
```
Comment (jingsong) [inline airflow/hooks/base_hook.py:48]: wonder if we should encode max_tries into ENV. we're grabbing the connection from ENV anyway? then default to 5 if we can't grab it from ENV.

File: airflow/settings.py [hunk 2]
```diff
@@ -183,6 +183,19 @@ def configure_orm(disable_connection_pool=False):
...
+    try:
+        connect_timeout_seconds = conf.getint('core', 'SQL_ALCHEMY_CONNECT_TIMEOUT_SECONDS')
+    except conf.AirflowConfigException:
+        connect_timeout_seconds = 10
+    engine_args['connect_timeout'] = connect_timeout_seconds
+
+    try:
+        statement_timeout_seconds = conf.getint('core', 'SQL_ALCHEMY_STATEMENT_TIMEOUT_SECONDS')
+    except conf.AirflowConfigException:
+        statement_timeout_seconds = 10
```
Comment (jingsong) [inline airflow/settings.py:196]: maybe statement timeout should be a bit bigger. sometimes we're reading a lot of state at once from postgres?

File: airflow/utils/db.py [hunk 3]
```diff
@@ -64,13 +65,19 @@ def wrapper(*args, **kwargs):
...
         session_in_kwargs = arg_session in kwargs
-
-        if session_in_kwargs or session_in_args:
-            return func(*args, **kwargs)
-        else:
-            with create_session() as session:
-                kwargs[arg_session] = session
-                return func(*args, **kwargs)
+        try_number = 1
+        while try_number < max_tries:
```
Comment (jingsong) [inline airflow/utils/db.py:69]: ```<=```

File: airflow/utils/db.py [hunk 4]
```diff
@@ -64,13 +65,19 @@ def wrapper(*args, **kwargs):
...
+            try:
+                if session_in_kwargs or session_in_args:
+                    return func(*args, **kwargs)
+                else:
+                    with create_session() as session:
+                        kwargs[arg_session] = session
+                        return func(*args, **kwargs)
+            except Exception as e:
+                log.warn(f'Try {try_number}/{max_tries} failed to perform db action.', e)
+                time.sleep(min(30., .01 * (1 << try_number)))
```
Comment (jingsong) [inline airflow/utils/db.py:79]: .01 * (1 << 5) = .32 seconds?

**Comment by jingsong** [comment]
i feel like we should pull the GCS client changes into a separate PR? we should roll these kind of changes more incrementally rather than lumped together. it will make diagnosing issues easier if/when they pop up

File: airflow/utils/db.py (hunk 4 above)
Comment (spennihana) [reply airflow/utils/db.py:79]: max_tries is set by the caller; max timeout never gets set, so cap it to 30s

File: airflow/settings.py (hunk 2 above)
Comment (spennihana) [reply airflow/settings.py:196]: this is a default; we can bump the default to something like 180?

Comment (spennihana) [reply airflow/settings.py:196]: The thinking is that this should always be fast; if it's not super fast then something might be super wrong. We should never be pulling out more than a few KB of data.

File: airflow/hooks/base_hook.py (hunk 1 above)
Comment (spennihana) [reply airflow/hooks/base_hook.py:48]: yep we can do that; i'll break out this from the gcs changes first

File: airflow/utils/db.py (hunk 4 above)
Comment (jingsong) [reply airflow/utils/db.py:79]: but if we're doing exponential backoff, then a .32 second sleep after 4 tries seems like it's not what we want?

Comment (spennihana) [reply airflow/utils/db.py:79]: nope, we want 0.1, not 0.01, typo'd the decimal

**Pull request closed without merging**